Issue/101 exp

fenruga opened the pull request:
#101

fenruga committed:
expAdd as command
status reworked
helpers

Commit:
setupChannelExp init
old exp deleted

fenruga committed:
updateExp reqorked
setupChannelExp command added
discord config suggestion

fenruga committed:
updateExp moved up in chain

fenruga committed:
comments removed

fenruga commented:
Здесь не хватает i18n
И я прям чувствую, как мне больно их писать каждый раз =\

fenruga commented on src/modules/commands/setupChannelExp.js:
```diff
@@ -0,0 +1,70 @@
+const command = require('../command.filter');
+const { discord: { channels } } = require('../../config');
+
+async function setDefaultsChannelExp({ updateModuleData }) {
```
Я не придумал куда это впихнуть по-хорошему? В `__INIT__`?

dcversus replied:
setDefaultsChannelExp а такого и не должно быть, все каналы мы сетапим вручную

fenruga replied:
Хз, кто-то из нас в вординг не смог снова =\\
Типа, как я понял, ты хотела какой-то предефайнд в конфиге, а остальное определять/переопределять через команду

dcversus replied:
Я хочу иметь возможность, сетапить некий конфиг без набора команд, в ОТНОШЕНИИ миссий итп. При чем тут экспа по каналам - хз.

fenruga replied:
'global settings' misleading

dcversus replied:
Не поняла комментария

dcversus commented on src/config.js:
```diff
@@ -13,5 +13,11 @@ module.exports = {
         userFields: ['username', 'id', 'discriminator', 'bot', 'avatar'],
+        channels: {
+            mainChannel: {
+                id: process.env.MAIN_CHANNEL_ID,
```
от этого надо избавится и обязательно должно быть что-то типо /setChannelExp amount

dcversus commented on src/modules/commands/expAdd.js:
```diff
@@ -0,0 +1,46 @@
+const isModerator = require('../isModerator');
+const command = require('../command.filter');
+const { getDiscordIdFromMention } = require('../helpers/discord');
+const { addUserExp } = require('../helpers/experience');
```
что-то мне не нравятся эти helpers блин, надо как-то грамотно обдумать, чтобы в этом не нуждаться
наверное, следует делать addUserExp именно модулем, я изначально тригеррилась, что addExp не САМОДОСТАТОЧЕН, и чет-то прокидывает и есть некий финалайзер... это было грязно. 

в идеале addUserExp это простой executor который смотрит `const {target, amount, reason, withMention} = request.args;` и если их видет то сразу же делает вызов к db и если есть withMention то отправляет в канал источник сообщение в котором спамит уведомлялкой

короч я про то, чтобы юзать напрямую модули и их как композицию

fenruga replied:
Снова вординг, но ок

dcversus commented on src/modules/commands/expAdd.js:
```diff
@@ -0,0 +1,46 @@
+
+const isModerator = require('../isModerator');
+const command = require('../command.filter');
+const { getDiscordIdFromMention } = require('../helpers/discord');
```
все тоже самое, это такой же модуль как command, который добавляет (парсит) input и кидает прям юзеров дальше в request

fenruga replied:
А вот здесь не вординг, но я не согласен. У нас сейчас 100% использования этой штуки не нужны "как модуль", а нужны "как функция". Прям здесь -- нам не нужны все mentions'ы распаршенные, нам нужен один позиционный аргумент распарсить

dcversus replied:
Композиция через модули, функций хелперов быть не может, а если и появляются и они напрямую работают с контекстом/реквестом, то они объявляются в init (как i18n)

dcversus replied:
Ну и чекнуть все mention не сложно, кстати там в message прям массив всех упоминаний есть

fenruga replied:
Я помню, что он там есть, но это не тот функционал, который нужен конкретно здесь
Не, я все равно не могу уложить в голове: нафига нам отдельный модуль для этого и чем он лучше функции. Абсолютно то же самое, связность такая же, юзабельность такая же

dcversus replied:
Мы не используем функции, у нас executor / module это first sitizen сущности. 

На самом деле, если у тебя простой хелпер, чистая функция которая просто берет простой аргумент и возвращает что-то новое (любая функция из lodash) то это допустимо. Если ты начинаешь прикидывать. req или ctx то значит уже что-то идёт не так.

Зачем getMentions отдельным модулем? Затем чтобы: а) не ебаться потом с интеграцией телеграмма и утуба (там они по другому обозначаются) и б) чтобы не парсить строку, а юзать msg.mentions и в) очевидно что часто требуется юзать mention, много кейсов где надо много их вытащить и желательно не как чистый аргумент, а как именно подстроку. 

Сложность не увеличивает, реюзабельность растет

dcversus commented on src/modules/commands/setupChannelExp.js:
```diff
@@ -0,0 +1,70 @@
+    } = ctx;
+    const {
+        from,
+        args: { amount },
```
я бы еще добавила бы reason (необязательный) для кастомизации событий с каналов типо "схватил 150 экспы когда писал в ГИЛЬДОВАЯ ФЛУДИЛЬНЯ"

dcversus commented on src/modules/commands/setupChannelExp.js:
```diff
@@ -0,0 +1,70 @@
+};
+
+module.exports = [
+    isModerator,
```
это бы писать на одной строчке с нижестоящим модулем и чисто по логике тут 
```suggestion
    isModerator, command('setupChannelExp amount'), setupChannelExp
```
это же одна цепочка, один набор условий и действий

dcversus commented on src/modules/helpers/discord.js:
```diff
@@ -0,0 +1,10 @@
+function getDiscordIdFromMention(mention) {
```
я уже писала, хелперы как функции - недопустимый метод организации кода, стараемся все делать через модули (а ежики плакали, но продолжали... кхм)

dcversus commented on src/modules/commands/expAdd.js:
```diff
@@ -0,0 +1,46 @@
+const isEmpty = require('lodash/isEmpty');
+
```
в идеале ВЕСЬ ЭТОТ МОДУЛЬ должен выглядеть как модуль в котором ровно одна строчка кода
```suggestion
module.exports = [isModerator, command('expAdd target amount reason'), addExp]
```

где addExp это универсальный модуль, который смотрит args на наличие указанных выше параметров, которые любой другой модуль может сетапить

fenruga replied:
Т.е. как это было сделано раньше...*see_no_evil*

dcversus replied:
Нет, раньше оно не смотрело request а просто каррировало количество опыта. Раньше оно занималось хуйней и просто обновляло некий request.exp. Сейчас должно сразу же обновлять опыт у юзера и смотреть аргументы из request (чтобы тот же command мог бы прозрачно передавать данные)

fenruga replied:
На самом деле, почему в args добавлять -- тоже не понятно, это ж для команд
В смысле сразу же добавлять, если "любой другой модуль может сетапить" и мы должны "смотреть аргументы из request"?
Как было: писалось хоть откуда `req.exp = someAmount`, в отдельном модуле обновлялось значение в бд

dcversus replied:
Модуль - самодостаточная единица, оптимизации на уровне, когда ты ещё не понимаешь что оптимизируешь - всратая идея

dcversus replied:
Насчёт args спорно. Но суть такая что один модуль устанавливает значения, другой их принимает и юзает. Для кейса - команда даёт опыт, юзать args норм. Если ты создашь хранилище в request отдельное, то у тебя будет модуль который из args просто перекладывает в другое место (object mapping) что тоже норм.

fenruga replied:
Ну, окей, т.е. вся разница со старым должна быть в том, чтобы вместо одного обновления делать обновления каждый раз
Оки)

dcversus commented on src/modules/updateExp.js:
```diff
@@ -1,84 +1,23 @@
-const extend = require('lodash/extend');
-const defaults = require('lodash/defaults');
+const { addUserExp } = require('./helpers/experience');
```
комментарий ко всему этому файлу

- название всратое и непонятное, должно быть как `addExpForChannelMessage`
- в идеале просто в args добавляет из db amount/reason/target
- типо также выглядит в две строчки (ну тут еще нужно сверить, есть ли вообще в db про этот channel... ну и вообще эти channel, речь на самом деле про from и должно поддерживаться универсально, как массив искомый в котором [adapter, ...]

fenruga replied:
1. Согласен
2. Ничего не понял. Совсем.
3. Все равно же зависимо от внешних модулей / того, как ты сетапишь from получается, не?

dcversus replied:
2) это типо, чтобы оно добавляло в request.args amount reason target данные и уже следом стоял модуль addExp (из обсуждения выше)

3) from это же просто уникальный ключ, можно делать toString целиком на него и будет збс как айди

fenruga replied:
3. Ну, сейчас это уникальный идентификатор сообщения, кажется

dcversus commented on src/modules/updateExp.js:
```diff
@@ -1,84 +1,23 @@
-        });
+    } = ctx;
+    const { channels } = await getModuleData('exp');
+    const currentChannel = from[1];
```
всрато, выше уже писала, надо целиком в базе from хранить и целиком from сравнивать

fenruga replied:
Да from в принципе весьма сам по себе всратый
Но да, имеет смысл, только не целиком
Сейчас хранится `{ channelId: { amount } }` я б тупо на `discord: { channelId: { amount } } }` поменял.

dcversus replied:
Зачем привязаться к discord, если можно атомарно хранить from как айди / ключ

dcversus commented:
по поводу i18n как ты думаешь было бы лучше писать их? что именно тебя смущает... надо подумать

dcversus committed:
Update src/modules/commands/setupChannelExp.js

Co-Authored-By: fenruga <fenrugaadv@gmail.com>

fenruga committed:
expAdd command refactor
addExpForChannelMessage defaults removed, only manual now
addUserExp module instead of helper

fenruga committed:
i18n added

fenruga committed:
Merge branch 'develop' of https://github.com/uz0/u.dcvs.ru into issue/101-exp
